cajita/src/Cajita_Halo.hpp

```diff
@@ -427,37 +509,512 @@ class Halo
             // Otherwise unpack the next buffer and apply the reduce operation.
             else
             {
-                auto subview =
-                    createSubview( view, _owned_spaces[unpack_index] );
-                auto owned_buffer = _owned_buffers[unpack_index];
-                IndexSpace<4> scatter_space(
-                    {static_cast<long>( subview.extent( 0 ) ),
-                     static_cast<long>( subview.extent( 1 ) ),
-                     static_cast<long>( subview.extent( 2 ) ),
-                     static_cast<long>( subview.extent( 3 ) )} );
-                Kokkos::parallel_for(
-                    "Cajita::Halo::scatterOwned",
-                    createExecutionPolicy( scatter_space, execution_space() ),
-                    KOKKOS_LAMBDA( const int i, const int j, const int k,
-                                   const int l ) {
-                        reduce( subview, owned_buffer, i, j, k, l );
-                    } );
+                unpackBuffer( reduce_op, exec_space,
+                              _owned_buffers[unpack_index],
+                              _owned_steering[unpack_index], arrays.view()... );
             }
 
             // Wait on send requests.
             MPI_Waitall( num_n, requests.data() + num_n, MPI_STATUSES_IGNORE );
         }
     }
 
+  public:
+    // Get the byte sizes of the array value types. Also check that the memory
+    // spaces are compatible.
+    template <std::size_t N, class Array_t>
+    void getByteSizes( std::array<std::size_t, N> &byte_sizes,
+                       const std::size_t array_idx, const Array_t & )
+    {
+        static_assert(
+            std::is_same<typename Array_t::memory_space, memory_space>::value,
+            "Memory space of array is different than halo" );
+        byte_sizes[array_idx] = sizeof( typename Array_t::value_type );
+    }
+
+    template <std::size_t N, class Array_t, class... ArrayTypes>
+    void getByteSizes( std::array<std::size_t, N> &byte_sizes,
+                       const std::size_t array_idx, const Array_t &,
+                       const ArrayTypes &... arrays )
```

Comment:
I really don't get why this need to be so convoluted.

First of, I think this overload never gets called.
Second if you must use a function for something that trivial (you only use it at one single place so you might as well inline the definition of the array) you could do
```cpp
template <class... ArrayTypes>
std::array<std::size_t, sizeof...(ArrayTypes)> getByteSizes( const ArrayTypes &... arrays )
{
    return {sizeof( typename ArrayTypes::value_type )...};
}

Reply:
good idea - I eliminated this function in favor of assignment.

Reply:
Was also able to eliminate the recursion that gets the shared index spaces.